nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp

```diff
@@ -134,6 +134,27 @@ class Costmap2DROS : public nav2_util::LifecycleNode
    */
   nav2_util::CallbackReturn on_shutdown(const rclcpp_lifecycle::State & state) override;
 
+  /**
+   * @brief override on_rcl_preshutdown() as empty
+   * [reason] costmap only could be created by its parents like planner/controller_server
```

Comment:
So, it would be useful to write a quick unit test to destroy the object https://github.com/ros-planning/navigation2/blob/main/nav2_util/src/lifecycle_node.cpp#L43-L54 and make sure the transition is also followed even if `rclcpp::shutdown()` isn't yet called

Reply:
> Incorrect, there is a costmap node `main.cpp` file that it can be launched independently

any code file show that?  we didn't find where the node is launched independently......

or you mean it just would be tested by launching independently ?

Reply:
> So, it would be useful to write a quick unit test to destroy the object 

absolutely, we find it failed in CI because the existing unit test launched it independently.

however, in our suggestion, we'd like to make it as a **child-LifecycleNode** fully dependent on some other lifecycle node.

Reply:
Its not currently used within the stack, but there are user that use it that way and it is a supported feature. You'd have to test this yourself

Reply:
> Its not currently used within the stack, but there are user that use it that way and it is a supported feature. 

ok, i get it. I will try to make this mechanism compatible with different usage modes, both as a child-LifecycleNode and as an independent LifecycleNode.

Reply:
- https://github.com/ros-planning/navigation2/blob/main/nav2_costmap_2d/src/costmap_2d_node.cpp

Like I said, its not run during nav2's default system, but it is used by people not only in testing but in production.